extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -0,0 +1,705 @@
+              streamPartition.getPartitionId()
+          );
+
+        }
+      } else {
+        log.info("Worker for partition[%s] has been stopped", streamPartition.getPartitionId());
+      }
+    }
+  }
+
+  // used for deaggregate
+  private final MethodHandle deaggregateHandle;
+  private final MethodHandle getDataHandle;
+
+  private final AmazonKinesis kinesis;
+
+  private final int recordsPerFetch;
+  private final int fetchDelayMillis;
+  private final boolean deaggregate;
+  private final int recordBufferOfferTimeout;
+  private final int recordBufferFullWait;
+  private final int fetchSequenceNumberTimeout;
+  private final int maxRecordsPerPoll;
+  private final int fetchThreads;
+  private final int recordBufferSize;
+
+  private ScheduledExecutorService scheduledExec;
+
+  private final Map<StreamPartition<String>, PartitionResource> partitionResources = new ConcurrentHashMap<>();
+  private BlockingQueue<OrderedPartitionableRecord<String, String>> records;
+
+  private volatile boolean checkPartitionsStarted = false;
+  private volatile boolean closed = false;
+
+  public KinesisRecordSupplier(
+      AmazonKinesis amazonKinesis,
+      int recordsPerFetch,
+      int fetchDelayMillis,
+      int fetchThreads,
+      boolean deaggregate,
+      int recordBufferSize,
+      int recordBufferOfferTimeout,
+      int recordBufferFullWait,
+      int fetchSequenceNumberTimeout,
+      int maxRecordsPerPoll
+  )
+  {
+    Preconditions.checkNotNull(amazonKinesis);
+    this.kinesis = amazonKinesis;
+    this.recordsPerFetch = recordsPerFetch;
+    this.fetchDelayMillis = fetchDelayMillis;
+    this.deaggregate = deaggregate;
+    this.recordBufferOfferTimeout = recordBufferOfferTimeout;
+    this.recordBufferFullWait = recordBufferFullWait;
+    this.fetchSequenceNumberTimeout = fetchSequenceNumberTimeout;
+    this.maxRecordsPerPoll = maxRecordsPerPoll;
+    this.fetchThreads = fetchThreads;
+    this.recordBufferSize = recordBufferSize;
+
+    /**
+     * the deaggregate function is implemented by the amazon-kinesis-client, whose license is not compatible with Apache.
+     * The work around here is to use reflection to find the deaggregate function in the classpath. See details on the
+     * docs page for more information on how to use deaggregation
+     */
+    if (deaggregate) {
+      try {
+        Class<?> kclUserRecordclass = Class.forName("com.amazonaws.services.kinesis.clientlibrary.types.UserRecord");
+        MethodHandles.Lookup lookup = MethodHandles.publicLookup();
+
+        Method deaggregateMethod = kclUserRecordclass.getMethod("deaggregate", List.class);
+        Method getDataMethod = kclUserRecordclass.getMethod("getData");
+
+        deaggregateHandle = lookup.unreflect(deaggregateMethod);
+        getDataHandle = lookup.unreflect(getDataMethod);
+      }
+      catch (ClassNotFoundException e) {
+        log.error(
```

Comment:
You don't have to print a log if you want to throw an exception. Please add the log message to the exception instead.